test/09-loki-backend.ts

```diff
@@ -0,0 +1,135 @@
+	it('should successfully publish log', async () => {
+		const loki = new LokiBackend();
+		const response = await loki.publish(createContext(), [createLog()]);
+		expect(response).to.be.not.null;
+	});
+
+	it('should store and retrieve device log', async () => {
+		const loki = new LokiBackend();
+		const ctx = createContext();
+		const log = createLog();
+		const response = await loki.publish(ctx, [_.clone(log)]);
+		expect(response).to.be.not.null;
+		const history = await loki.history(ctx, 1000);
+		expect(history[history.length - 1]).to.deep.equal(log);
+	});
+
+	it('should convert multiple logs with different labels to streams and then back to logs', function () {
+		const loki = new LokiBackend();
+		const ctx = createContext();
+		const logs = [
+			createLog(),
+			createLog(),
+			createLog({ serviceId: 1 }),
+			createLog({ serviceId: 2 }),
+			createLog({ serviceId: 3 }),
+		];
+		// @ts-ignore usage of private function
+		const streams = loki.fromDeviceLogsToStreams(ctx, _.cloneDeep(logs));
+		expect(streams.length).to.be.equal(
+			4,
+			'should be 4 streams since 5 logs have 4 distinct services (null, 1, 2, 3)',
+		);
+		// @ts-ignore usage of private function
+		const logsFromStreams = loki.fromStreamsToDeviceLogs(streams);
+		expect(logsFromStreams).to.deep.equal(logs);
+	});
+
+	it('should push multiple logs with different labels and return in order', async function () {
+		const loki = new LokiBackend();
+		const ctx = createContext();
+		const logs = [
+			createLog({ timestamp: Date.now() - 4 }),
+			createLog({ timestamp: Date.now() - 3 }),
+			createLog({ timestamp: Date.now() - 2, isStdErr: false }),
+			createLog({ timestamp: Date.now() - 1, isStdErr: false }),
+			createLog({ timestamp: Date.now(), isStdErr: false, isSystem: false }),
+		];
+		const response = await loki.publish(ctx, _.cloneDeep(logs));
+		expect(response).to.be.not.null;
+		const history = await loki.history(ctx, 1000);
+		expect(history.slice(-5)).to.deep.equal(logs);
+	});
+
+	it('should de-duplicate multiple identical logs', async function () {
+		const loki = new LokiBackend();
+		const ctx = createContext();
+		const log = createLog();
+		const logs = [_.clone(log), _.clone(log), _.clone(log)];
+		const response = await loki.publish(ctx, _.cloneDeep(logs));
+		expect(response).to.be.not.null;
+		const history = await loki.history(ctx, 1000);
+		expect(history[1].timestamp).to.not.equal(log.timestamp);
+	});
+
+	// TODO: test is flakey
+	it.skip('should subscribe and receive a published logs', async function () {
+		const ctx = createContext();
+		const loki = new LokiBackend();
+		return new Bluebird(async (resolve, reject) => {
+			loki.subscribe(ctx, (incomingLog) => {
+				try {
+					expect(incomingLog).to.deep.equal(log);
+					resolve();
+				} catch (err) {
+					reject(err);
+				}
+			});
+			const log = createLog();
+			loki.publish(ctx, [_.clone(log)]);
+		}).timeout(5000, 'Subscription did not receive log');
```

Comment:
This would be clearer imo as
```suggestion
		const log = createLog();
		const incomingLog = await new Bluebird(async (resolve) => {
			loki.subscribe(ctx, resolve);
			loki.publish(ctx, [_.clone(log)]);
		}).timeout(5000, 'Subscription did not receive log');
		expect(incomingLog).to.deep.equal(log);
```

Reply:
Good point. much nicer. updated.